{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import math\n",
    "import numpy as np\n",
    "import scipy\n",
    "from scipy import stats\n",
    "from scipy.stats import ranksums\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import anndata\n",
    "import warnings\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "# from concurrent.futures import ThreadPoolExecutor\n",
    "# from concurrent.futures import ProcessPoolExecutor\n",
    "from joblib import Parallel, delayed\n",
    "# import mplcursors\n",
    "# import matplotlib.cm as cm\n",
    "from multiprocessing import Process,Manager\n",
    "import dill as pickle\n",
    "\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SCPerturbationAnalysis:\n",
    "\n",
    "    def __init__(self, file_path):\n",
    "        self.__anndata = sc.read_h5ad(file_path)\n",
    "        self.__expression = pd.DataFrame(self.__anndata.X, columns=self.__anndata.var.index, index=self.__anndata.obs.index).T\n",
    "        # self.__perturb_fliter,self.__perturbation = self._fliter_by_cell_num()\n",
    "        self.__perturbation=list(set(self.__anndata.obs['gene']))\n",
    "        self.__gene_exp = self._calculate_gene_exp()\n",
    "        self.__lfc = self._calculate_lfc()\n",
    "        self.__pvalue = self._calculate_gene_pvalue()\n",
    "        # self.__total_mRNA = self._calculate_total_mRNA()\n",
    "        # self.__knock_gene_df = self._calculate_knocked_gene_info()\n",
    "        # self.__downstream_counts, self.__downstream_genelist=self._find_downstream_gene()\n",
    "        # self.__uptream_counts, self.__upstream_genelist=self._find_upstream_gene()\n",
    "\n",
    "    def _fliter_by_cell_num(self):\n",
    "        ### generate cell number distribution\n",
    "        perturbation=list(set(self.__anndata.obs['gene']))\n",
    "        cell_num=[]\n",
    "        for gene in perturbation:\n",
    "            expriment=self.__expression[self.__anndata.obs[self.__anndata.obs['gene']==gene].index]\n",
    "            cell_num.append(expriment.shape[1])\n",
    "        cell_num=pd.DataFrame(cell_num,index=perturbation,columns=['cell_num'])\n",
    "\n",
    "        total_rna_std_log = [np.std(np.log2(self.__anndata.obs[self.__anndata.obs['gene'] == gene]['UMI_count'])) for gene in perturbation]\n",
    "        total_rna_log = [np.log2(self.__anndata.obs[self.__anndata.obs['gene'] == gene]['UMI_count']).mean() for gene in perturbation]\n",
    "        total_rna_std=np.exp(total_rna_std_log)\n",
    "        total_rna = np.exp(total_rna_log)\n",
    "        perturb_fliter=pd.DataFrame(zip(total_rna,total_rna_std,total_rna_std/total_rna),index=perturbation,columns=['total_rna','total_rna_std','total_rna_cv'])\n",
    "        perturb_fliter['cell_number']=cell_num['cell_num']\n",
    "\n",
    "        ### update perturbation list after fliter with cell_number\n",
    "        pertubation_flited=perturb_fliter[perturb_fliter['cell_number']>15].index\n",
    "\n",
    "        return perturb_fliter,pertubation_flited\n",
    "\n",
    "    def _calculate_gene_exp(self):\n",
    "        gene_exp = pd.DataFrame()\n",
    "        for gene in self.__perturbation:\n",
    "            cells = self.__anndata.obs[self.__anndata.obs['gene'] == gene]\n",
    "            expression_filtered = pd.DataFrame(self.__expression[cells.index])\n",
    "            gene_exp = gene_exp.append(expression_filtered.mean(axis=1), ignore_index=True)\n",
    "        gene_exp.index = self.__perturbation\n",
    "        return gene_exp\n",
    "\n",
    "    def _calculate_lfc(self):\n",
    "        lfc = self.__gene_exp / self.__gene_exp.loc['non-targeting']\n",
    "        lfc = np.log2(lfc.replace([np.inf, -np.inf], [1000, -1000]))\n",
    "        rename_dict = dict(zip(self.__anndata.var.index, self.__anndata.var['gene_name']))\n",
    "        rename_dict['ENSG00000284024'] = 'MSANTD7'\n",
    "        lfc.rename(columns=rename_dict, inplace=True)\n",
    "        return lfc\n",
    "\n",
    "    def _calculate_gene_pvalue(self):\n",
    "        non_targeting = self.__expression[self.__anndata.obs[self.__anndata.obs['gene'] == 'non-targeting'].index]\n",
    "\n",
    "        # Function to calculate p-value for a gene\n",
    "        def ranksums_parallel(gene):\n",
    "            multitest = False\n",
    "            cells = self.__anndata.obs[self.__anndata.obs['gene'] == gene]\n",
    "            expression_filtered = self.__expression[cells.index]\n",
    "            p_values = expression_filtered.apply(lambda row: ranksums(row, non_targeting.loc[row.name])[1], axis=1)\n",
    "            if multitest:\n",
    "                p_values = multipletests(p_values, method='fdr_bh')[1]\n",
    "            return p_values\n",
    "\n",
    "        # Use Parallel to calculate p-values for all genes in parallel\n",
    "        num_cores = -1  # Use all available cores, adjust as needed\n",
    "        with Parallel(n_jobs=num_cores) as parallel:\n",
    "            results = parallel(delayed(ranksums_parallel)(gene) for gene in self.__perturbation)\n",
    "\n",
    "        gene_pvalue = pd.DataFrame({gene: p_values for gene, p_values in zip(self.__perturbation, results)})\n",
    "        rename_dict = dict(zip(self.__anndata.var.index, self.__anndata.var['gene_name']))\n",
    "        rename_dict['ENSG00000284024'] = 'MSANTD7'\n",
    "        gene_pvalue.rename(index=rename_dict, inplace=True)\n",
    "        gene_pvalue = gene_pvalue.T\n",
    "        return gene_pvalue\n",
    "\n",
    "    def _calculate_total_mRNA(self):\n",
    "        control = self.__anndata.obs[self.__anndata.obs['gene'] == 'non-targeting']\n",
    "        total_rna_log = [np.log2(self.__anndata.obs[self.__anndata.obs['gene'] == gene]['UMI_count']).mean() for gene in self.__perturbation]\n",
    "        p_value = [scipy.stats.ttest_ind(np.log2(control['UMI_count']), np.log2(self.__anndata.obs[self.__anndata.obs['gene'] == gene]['UMI_count']))[1] for gene in self.__perturbation]\n",
    "        adjusted_pvalue=multipletests(p_value,method='fdr_bh')[1]\n",
    "        total_rna = np.exp(total_rna_log)\n",
    "        total_mRNA = pd.DataFrame(total_rna, index=self.__perturbation, columns=['total_mRNA'])\n",
    "        total_mRNA['ratio'] = total_mRNA / total_mRNA.loc['non-targeting', 'total_mRNA']\n",
    "        total_mRNA['logFC'] = np.log2(total_mRNA['ratio'])\n",
    "        total_mRNA['p_value'] = adjusted_pvalue\n",
    "        return total_mRNA\n",
    "\n",
    "    def _calculate_knocked_gene_info(self):\n",
    "        gene_name_ary = []\n",
    "        lfc_ary = []\n",
    "        gene_ary_no_knock = []\n",
    "        gene_ary_knocked = []\n",
    "        rename_dict = dict(zip(self.__anndata.var.index, self.__anndata.var['gene_name']))\n",
    "        rename_dict['ENSG00000284024'] = 'MSANTD7'\n",
    "        self.__gene_exp.rename(columns=rename_dict, inplace=True)\n",
    "        for knock_gene in self.__lfc.index:\n",
    "            if knock_gene in self.__lfc.columns:\n",
    "                gene_name_ary.append(knock_gene)\n",
    "                lfc_ary.append(self.__lfc.loc[knock_gene, knock_gene])\n",
    "                gene_ary_knocked.append(self.__gene_exp.loc[knock_gene, knock_gene])\n",
    "                gene_ary_no_knock.append(self.__gene_exp.loc['non-targeting', knock_gene])\n",
    "        knock_gene_df = pd.DataFrame(zip(gene_name_ary, lfc_ary, gene_ary_no_knock, gene_ary_knocked), columns=['gene_name', 'log FC', 'original', 'knocked'])\n",
    "        knock_gene_df.set_index(\"gene_name\", inplace=True)\n",
    "        knock_gene_df['variation']=knock_gene_df['knocked']-knock_gene_df['original']\n",
    "        ### 计算被敲除基因的相对值\n",
    "        filtered = self.__total_mRNA[self.__total_mRNA.index.isin(knock_gene_df.index)]\n",
    "        knock_gene_df['original_relative'] = knock_gene_df['original'] / self.__total_mRNA.loc['non-targeting', 'total_mRNA']\n",
    "        knock_gene_df['knocked_relative'] = knock_gene_df['knocked'] / filtered['total_mRNA']\n",
    "        knock_gene_df['variation_relative'] = knock_gene_df['knocked_relative'] - knock_gene_df['original_relative']\n",
    "        return knock_gene_df\n",
    "\n",
    "    def _find_downstream_gene(self):\n",
    "        # Calculate the absolute difference between knocked and non-targeting for all pairs (i, j)\n",
    "        # Pre-calculate the threshold for all rows\n",
    "        non_targeting = self.__gene_exp.loc['non-targeting']\n",
    "        thresholds = pd.DataFrame(np.percentile(np.abs(self.__gene_exp - non_targeting), 80, axis=1),index=self.__lfc.index)\n",
    "        genelist = []\n",
    "\n",
    "        for i in self.__lfc.index:\n",
    "            abs_diff_non_targeting = np.abs(non_targeting - self.__gene_exp.loc[i])\n",
    "            data=np.full((len(abs_diff_non_targeting),1),thresholds.loc[i])\n",
    "            diff=pd.DataFrame(data,index=abs_diff_non_targeting.index)\n",
    "            #\n",
    "            mask = ((np.abs(self.__lfc.loc[i]) > 1.5).values)&(abs_diff_non_targeting > diff.iloc[:,0]) & ((p_value.loc[i].values < 0.05))\n",
    "            gene_list = self.__lfc.columns[mask].tolist()\n",
    "            genelist.append(gene_list)\n",
    "       \n",
    "        downstream_gene_list={key: values for key,values in zip(self.__lfc.index, genelist)}\n",
    "        downstream_gene_list={key: [item for item in value if item!=key] for key,value in downstream_gene_list.items()}\n",
    "        downstream_num=[]\n",
    "        for key,value in downstream_gene_list.items():\n",
    "            downstream_num.append(len(value))\n",
    "        downstream_gene_counts=pd.DataFrame(downstream_num,index=lfc.index,columns=['counts'])\n",
    "        return downstream_gene_counts, downstream_gene_list\n",
    "    \n",
    "    def _find_upstream_gene(self):\n",
    "        upstream_dict = {element: [] for element in self.__lfc.columns}\n",
    "        for key in self.__downstream_genelist:\n",
    "            for element in self.__lfc.columns:\n",
    "                if element in self.__downstream_genelist[key]:\n",
    "                    upstream_dict[element].append(key)\n",
    "        upstream_num=[]\n",
    "        for key,value in upstream_dict.items():\n",
    "            upstream_num.append(len(value))\n",
    "        upstream_num=pd.DataFrame(upstream_num,index=self.__lfc.columns,columns=['counts'])\n",
    "        return upstream_num, upstream_dict\n",
    "\n",
    "    # def scatter_vocano_for_gene(self, gene, output_expression_is_0):\n",
    "    #     expriment=self.__expression[self.__anndata.obs[self.__anndata.obs['gene']==gene].index]\n",
    "    #     control=self.__expression[self.__anndata.obs[self.__anndata.obs['gene']=='non-targeting'].index].mean(axis=1)\n",
    "    #     control.index=self.__lfc.columns\n",
    "\n",
    "    #     # %matplotlib\n",
    "    #     # control=self.__self.__expression[self.__self.__self.__anndata.obs[self.__self.__self.__anndata.obs['gene']=='non-targeting'].index].sample(n=expriment.shape[1],axis=1).mean(axis=1)\n",
    "    #     expriment1=expriment.mean(axis=1)\n",
    "    #     expriment1.index=self.__lfc.columns\n",
    "    #     # print(f\"knocked gene\", gene)\n",
    "    #     # print(f\"cell number:\",expriment.shape[1])\n",
    "\n",
    "    #     ### scatter plot and valcano\n",
    "    #     fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12,6))\n",
    "    #     plt.rcParams.update({'font.size': 10})\n",
    "    #     sct=ax1.scatter(np.log2(control),(np.log2(expriment1)),s=10)\n",
    "    #     downstream_genes=self.__downstream_genelist[gene]\n",
    "    #     ax1.scatter(np.log2(control[control.index.isin(downstream_genes)]),(np.log2(expriment1[control.index.isin(downstream_genes)])),edgecolors='orange',s=10, label='downstream genes')\n",
    "    #     ax1.scatter(np.log2(control[gene]),(np.log2(expriment1[gene])),color='red',s=10, label='downstream genes')\n",
    "    #     cursor = mplcursors.cursor(sct, hover=True)\n",
    "    #     # by default the annotation displays the xy positions\n",
    "    #     @cursor.connect(\"add\")\n",
    "    #     def on_add(sel):\n",
    "    #         sel.annotation.set_text([control.index[sel.index],self.__gene_exp.loc[gene,control.index[sel.index]]\n",
    "    #                                 ,self.__gene_exp.loc['non-targeting',control.index[sel.index]]])\n",
    "\n",
    "    #     ax1.plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control)),np.log2(max(control))],color='red',linestyle='--',label='y=x')\n",
    "    #     ratio=self.__total_mRNA.loc[gene,'ratio']\n",
    "    #     ax1.plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control))*ratio,np.log2(max(control))*ratio],color='red',label=f\"total_mRNS_ratio={ratio:.2f}\")\n",
    "    #     ax1.set_xlabel('log2(original UMI count)')\n",
    "    #     ax1.set_ylabel('log2(perturbed UMI count)')\n",
    "    #     ax1.set_title('original VS knocked-down gene expression')\n",
    "    #     ax1.legend(loc='upper left',bbox_to_anchor=(0,1), frameon=False)\n",
    "\n",
    "    #     p_values=self.___pvalue.loc[gene]\n",
    "    #     log2_fold_change = self.___lfc.loc[gene]\n",
    "    #     data = pd.DataFrame({'log2_fold_change': log2_fold_change, 'p_value': p_values},index=self.__gene_lfc.columns)\n",
    "    #     p_value_threshold = 0.05\n",
    "    #     fold_change_threshold = 1.5\n",
    "    #     voc=ax2.scatter(data['log2_fold_change'], -np.log10(data['p_value']),color='gray', alpha=0.7)\n",
    "    #     plt.scatter(data[data.index.isin(downstream_genes)]['log2_fold_change'], -np.log10(data[data.index.isin(downstream_genes)]['p_value']), color='red', alpha=0.7)\n",
    "    #     cursor = mplcursors.cursor(voc, hover=True)\n",
    "    #     @cursor.connect(\"add\")\n",
    "    #     def on_add(sel):\n",
    "    #         sel.annotation.set_text(data.index[sel.index])\n",
    "    #     ax2.axhline(-np.log10(p_value_threshold), color='blue', linestyle='--', label='p-value threshold')\n",
    "    #     ax2.axvline(fold_change_threshold, color='green', linestyle='--', label='fold change threshold')\n",
    "    #     ax2.axvline(-fold_change_threshold, color='green', linestyle='--')\n",
    "    #     ax2.set_xlabel('log2 Fold Change')\n",
    "    #     ax2.set_ylabel('-log10(p-value)')\n",
    "    #     ax2.set_title('Gene Volcano Plot')\n",
    "    #     if output_expression_is_0==True:\n",
    "    #         expr=pd.DataFrame(expriment1,columns=['UMI'])\n",
    "    #         zero_gene=expr[expr['UMI']==0].index\n",
    "    #         print(f'genes with 0 expression after perturbation)')\n",
    "    #         return zero_gene\n",
    "\n",
    "    @property\n",
    "    def gene_exp(self):\n",
    "        gene_exp=self.__gene_exp\n",
    "        gene_exp.columns=self.__lfc.columns\n",
    "        return gene_exp\n",
    "    \n",
    "    @property\n",
    "    def perturb_fliter(self):\n",
    "        return self.__perturb_fliter\n",
    "\n",
    "    @property\n",
    "    def gene_lfc(self):\n",
    "        return self.__lfc\n",
    "\n",
    "    @property\n",
    "    def gene_pvalue(self):\n",
    "        return self.__pvalue\n",
    "\n",
    "    @property\n",
    "    def total_mRNA_info(self):\n",
    "        return self.__total_mRNA\n",
    "\n",
    "    @property\n",
    "    def knocked_gene_info(self):\n",
    "        return self.__knock_gene_df\n",
    "\n",
    "    @property\n",
    "    def downstream_gene_counts(self):\n",
    "        return self.__downstream_counts\n",
    "    \n",
    "    @property\n",
    "    def downstream_gene_list(self):\n",
    "        return self.__downstream_genelist   \n",
    "    \n",
    "    @property\n",
    "    def uptream_gene_counts(self):\n",
    "        return self.__uptream_counts\n",
    "    \n",
    "    @property\n",
    "    def upstream_gene_list(self):\n",
    "        return self.__upstream_genelist  \n",
    "\n",
    "\n",
    "\n",
    "    def get_all_info_for_perturbation(self,gene):\n",
    "        output_df1=pd.DataFrame(zip(self.__gene_exp.loc[gene].T, self.__gene_exp.loc['non-targeting'].T, self.__lfc.loc[gene].T, self.__pvalue.loc[gene]),columns=['gene_exp','gene_exp_control','lfc','gene_pvalue'], index=self.__lfc.columns).T\n",
    "        print(f\"knocked gene is {gene}, and the basic information is shown as below: \")\n",
    "        print(output_df1[gene].to_string())\n",
    "        print()\n",
    "        print(f\"cell number of this perturbation:\")\n",
    "        print(self.__perturb_fliter.loc[gene,'cell_number'])\n",
    "        print()\n",
    "        print(f\"total_mRNA info:\")\n",
    "        print(self.__total_mRNA.loc[gene])\n",
    "        print()\n",
    "        print(f\"knocked gene info:\")\n",
    "        print(self.__knock_gene_df.loc[gene])\n",
    "        print()\n",
    "        print(f\"downstream genes counts is: {self.__downstream_counts.loc[gene,'counts']}\")\n",
    "        print(f\"downstream genes:\")\n",
    "        print(self.__downstream_genelist[gene])\n",
    "        print()\n",
    "        print(f\"upstream genes counts is: {self.__uptream_counts.loc[gene,'counts']}\")\n",
    "        print(f\"upstream genes:\")\n",
    "        print(self.__upstream_genelist[gene])\n",
    "\n",
    "        self.scatter_vocano_for_gene(gene,True)\n",
    "\n",
    "        \n",
    "        ### 对于要观测的target gene，观测其下游基因的表达分布及该perturbation在分布中的位置\n",
    "        print(f\"Gene expression level for each downstream gene\")\n",
    "        gene_list = list(self.__downstream_genelist[gene])\n",
    "        num_genes = len(gene_list)\n",
    "        cols = 6\n",
    "        rows = (num_genes + cols - 1) // cols  # Round up to the nearest integer\n",
    "        plt.figure(dpi=300)\n",
    "        gs = plt.GridSpec(rows, cols)  # Create a grid for subplots\n",
    "        plt.rcParams['xtick.direction'] = 'in'\n",
    "        plt.rcParams['ytick.direction'] = 'in'\n",
    "        plt.rcParams.update({'font.size': 3})\n",
    "        for i, gene_name in enumerate(gene_list):\n",
    "            ax = plt.subplot(gs[i])  # Create a subplot in the grid\n",
    "            data = self.__gene_exp[gene_name]\n",
    "            sns.distplot(data, axlabel=gene_name, ax=ax)\n",
    "            ax.axvline(self.__gene_exp.loc[gene, gene_name], color='red', lw=0.5)\n",
    "        # Remove any empty subplots in the grid\n",
    "        for i in range(num_genes, rows * cols):\n",
    "            plt.delaxes(plt.subplot(gs[i]))\n",
    "        plt.subplots_adjust(hspace=0.5, wspace=0.5)\n",
    "        plt.tick_params(pad=0.1)\n",
    "        plt.show()\n",
    "\n",
    "        ### sequencing quality\n",
    "        # df=self.__expression[self.__anndata.obs[self.__anndata.obs['gene']==gene].index]\n",
    "        # zero_per=df.eq(0).sum()/df.shape[0]\n",
    "        # print(f\"mean zero_percent:\",zero_per.mean())\n",
    "        # print(f\"Sequencing quality:\")\n",
    "        # plt.rcParams.update({'font.size':10})\n",
    "        # sns.displot(zero_per)\n",
    "        # plt.xlabel(\"zero percent of each cell\")\n",
    "        # plt.show()\n",
    "        # print()\n",
    "\n",
    "\n",
    "        # ### 画火山图\n",
    "        # print(f\"火山图：\")\n",
    "        # plt.rcParams.update({'font.size':10})\n",
    "        # p_values=self.__pvalue.loc[gene]\n",
    "        # log2_fold_change = self.__lfc.loc[gene]\n",
    "        # data = pd.DataFrame({'log2_fold_change': log2_fold_change, 'p_value': p_values},index=self.__lfc.index)\n",
    "        # p_value_threshold = 0.05\n",
    "        # fold_change_threshold = 1.5\n",
    "        # plt.scatter(data['log2_fold_change'], -np.log10(data['p_value']), color='gray', alpha=0.7)\n",
    "        # down_list=self.__downstream_genelist[gene]\n",
    "        # plt.scatter(data[data.index.isin(down_list)]['log2_fold_change'], -np.log10(data[data.index.isin(down_list)]['p_value']), color='red', alpha=0.7)\n",
    "        # plt.axhline(-np.log10(p_value_threshold), color='blue', linestyle='--', label='p-value threshold')\n",
    "        # plt.axvline(fold_change_threshold, color='green', linestyle='--', label='fold change threshold')\n",
    "        # plt.axvline(-fold_change_threshold, color='green', linestyle='--')\n",
    "        # plt.xlabel('log2 Fold Change')\n",
    "        # plt.ylabel('-log10(p-value)')\n",
    "        # plt.legend()\n",
    "        # plt.title('Volcano Plot')\n",
    "        # plt.show()\n",
    "        # print()\n",
    "\n",
    "        ### 画total mRNA的分布图\n",
    "        # print(f\"total mRNA distribution:\")\n",
    "        # sns.distplot(self.__total_mRNA['total_mRNA'])\n",
    "        # plt.axvline(self.__total_mRNA.loc[gene,'total_mRNA'], color='red', label=gene)\n",
    "        # plt.show()\n",
    "        # print()\n",
    "\n",
    "        ### 画下游基因counts分布图\n",
    "        # print(f\"log2(downstream genes count) distribution:\")\n",
    "        # sns.distplot(np.log2(self.__downstream_counts['counts']+1))\n",
    "        # plt.axvline(np.log2(self.__downstream_counts.loc[gene,'counts']+1), color='red', label=gene)\n",
    "        # plt.xlabel('log2(downstream genes count)')\n",
    "        # plt.show()\n",
    "        # print()\n",
    "\n",
    "\n",
    "        # ### output cell number and scatter(expression of original and perturbed together with total mRNA ratio line)\n",
    "        # expriment=self.__expression[self.__anndata.obs[self.__anndata.obs['gene']==gene].index]\n",
    "        # control=self.__expression[self.__anndata.obs[self.__anndata.obs['gene']=='non-targeting'].index].mean(axis=1)\n",
    "        # expriment1=expriment.mean(axis=1)\n",
    "        # print(f\"knocked gene\", gene)\n",
    "        # print(f\"cell number:\",expriment.shape[1])\n",
    "\n",
    "        # sns.scatterplot(np.log2(control),(np.log2(expriment1)))\n",
    "        # plt.plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control)),np.log2(max(control))],color='red',linestyle='--',label='y=x')\n",
    "        # ratio=self.__total_mRNA.loc[gene,'ratio']\n",
    "        # plt.plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control))*ratio,np.log2(max(control))*ratio],color='red',label=f\"total_mRNS_ratio={ratio:.2f}\")\n",
    "        # plt.xlabel('log2(original gene count)')\n",
    "        # plt.ylabel('log2(perturbed gene count)')\n",
    "        # plt.legend()\n",
    "        # print()\n",
    "\n",
    "    def generate_global_stats(self):\n",
    "        \n",
    "        # ### sequencing quality\n",
    "        # print(f\"Overview of sequencing quality:\")\n",
    "        # zero_per=self.__expression.eq(0).sum()/self.__expression.shape[0]\n",
    "        # plt.rcParams.update({'font.size':10})\n",
    "        # sns.displot(zero_per)\n",
    "        # plt.xlabel(\"zero percent of each cell\")\n",
    "        # print(f\"mean zero_percent for all cell:\",zero_per.mean())\n",
    "        # plt.show()\n",
    "        \n",
    "        ### cell number distribution\n",
    "        print(f\"cell number distribution for each perturbation:\")\n",
    "        %matplotlib inline\n",
    "        sns.distplot(np.log2(self.__perturb_fliter['cell_number']))\n",
    "        plt.axvline(x=np.log2(self.__perturb_fliter.loc['non-targeting','cell_number']), color='red', label='non-targeting')\n",
    "        plt.xlabel('log2(cell_number)')\n",
    "        plt.title('Perturtion cell number distribution')\n",
    "        plt.legend()\n",
    "        plt.show()\n",
    "        print()\n",
    "\n",
    "        ### plotg scatter: cell_number--total_mrna_cv  (the basis of filter accroding to cell_number)\n",
    "        print(f\"relationship between cell number and total mRNA CV:\")\n",
    "        sns.scatterplot(np.log2(self.__perturb_fliter['cell_number']),self.__perturb_fliter['total_rna_cv'])\n",
    "        plt.axvline(x=4, color='red', label='threshold')\n",
    "        plt.xlabel('log2(cell_number)')\n",
    "        plt.legend()\n",
    "        plt.show()\n",
    "        print()\n",
    "\n",
    "        ### total mRNA distribution visualization\n",
    "        print(f\"total mRNA distribution:\")\n",
    "        sns.distplot(self.__total_mRNA['total_mRNA'])\n",
    "        plt.axvline(self.__total_mRNA.loc['non-targeting','total_mRNA'], color='red', label='non-targeting')\n",
    "        plt.legend()\n",
    "        plt.show()\n",
    "\n",
    "        ### 画total mRNA火山图\n",
    "        # print(f\"total mRNA volcano plot:\")\n",
    "        # p_values=self.__total_mRNA['p_value']\n",
    "        # log2_fold_change = self.__total_mRNA['logFC']\n",
    "        # data = pd.DataFrame({'log2_fold_change': log2_fold_change, 'p_value': p_values})\n",
    "        # p_value_threshold = 0.05\n",
    "        # fold_change_threshold = 1\n",
    "        # plt.scatter(data['log2_fold_change'], -np.log10(data['p_value']), alpha=0.7)\n",
    "        # plt.xlabel('log2 Fold Change')\n",
    "        # plt.ylabel('-log10(p-value)')\n",
    "        # plt.title('total mRNA Volcano Plot')\n",
    "        # plt.show()\n",
    "\n",
    "        ### scatter: knocked gene change and total mRNA\n",
    "        %matplotlib\n",
    "        total_mRNA_hue=self.__total_mRNA[self.__total_mRNA.index.isin(self.__knock_gene_df.index)]\n",
    "        fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12,6), gridspec_kw={'width_ratios': [3, 2]})\n",
    "        ax1.tick_params(axis='both',)\n",
    "        ax1.set_title('original VS knocked-down gene expression level comparison')\n",
    "        plt.rcParams.update({'font.size': 10})\n",
    "        # scatter plot\n",
    "        sc = ax1.scatter(np.log2(self.__knock_gene_df['original']),np.log2(self.__knock_gene_df['knocked']),c=total_mRNA_hue['ratio'].rename('total mRNA ratio'),cmap='bwr')\n",
    "        ax1.plot([-9,max(np.log2(self.__knock_gene_df['original']))],[-9,max(np.log2(self.__knock_gene_df['original']))],color='gray',linestyle='--')\n",
    "        cbar=plt.colorbar(sc)\n",
    "        ax1.set_xlabel('log2(original gene expression)')\n",
    "        ax1.set_ylabel('log2(knocked-down gene experssion)')\n",
    "        ax2.set_xticks([])\n",
    "        ax2.set_yticks([])\n",
    "        ax2.set_title('Detail info')\n",
    "        ax2.set_xlim(0,2)\n",
    "        cursor = mplcursors.cursor(sc, hover=True)# by default the annotation displays the xy positions\n",
    "        @cursor.connect(\"add\")\n",
    "        def on_add(sel):\n",
    "            sel.annotation.set_text(self.__knock_gene_df.index[sel.index])\n",
    "            for text in ax2.texts:\n",
    "                text.set_visible(False)\n",
    "            source =self.__knock_gene_df.index[sel.index]\n",
    "            # ax2.text(0.03, 0.2, source, ha='left', va='top', fontsize=7)\n",
    "            ax2.axhline(0.3,lw=0.2,color='black',linestyle='--')\n",
    "            ax2.text(0.03,0.9,\"original expresison:\")\n",
    "            ax2.text(2,0.9,self.__gene_exp.loc['non-targeting',self.__knock_gene_df.index[sel.index]],ha='right')\n",
    "            ax2.text(0.03,0.8,\"knocked expresison:\")\n",
    "            ax2.text(2,0.8,self.__gene_exp.loc[self.__knock_gene_df.index[sel.index],self.__knock_gene_df.index[sel.index]],ha='right')\n",
    "            ax2.text(0.03,0.7,\"original total mRNA:\")\n",
    "            ax2.text(2,0.67,self.__total_mRNA.loc['non-targeting','total_mRNA'],ha='right')\n",
    "            ax2.text(0.03,0.6,\"knocked total mRNA:\")\n",
    "            ax2.text(2,0.57,self.__total_mRNA.loc[self.__knock_gene_df.index[sel.index],'total_mRNA'],ha='right')  \n",
    "            ax2.text(0.03,0.5,\"amount of downstream genes:\")\n",
    "            ax2.text(2,0.5,self.__downstream_counts.loc[self.__knock_gene_df.index[sel.index],'counts'],ha='right')\n",
    "            ax2.text(0.03,0.4,\"amount of upstream genes:\")\n",
    "            ax2.text(2,0.4,self.__uptream_counts.iloc[sel.index,0],ha='right')\n",
    "            ax2.text(1,0.25,\"Colors in scatterplot: total mRNA ratio\",ha='center')\n",
    "            ax2.text(1,0.15,\"For more detailed infomation,\",ha='center')\n",
    "            ax2.text(1,0.12,\"please run:\",ha='center')\n",
    "            ax2.text(1,0.09,\"_.get_all_info_for_perturbation(gene)\",ha='center')\n",
    "        plt.legend()\n",
    "        plt.show()\n",
    "        print()\n",
    "\n",
    "    \n",
    "        ### info of knocked gene\n",
    "        print(f\"Details of the knocked-down genes:\")\n",
    "        print(f\"All genes info higher expressed after knocked-down:\")\n",
    "        abnormal=self.__knock_gene_df[self.__knock_gene_df['variation']>0]\n",
    "        for gene in abnormal.index:\n",
    "            self.observe_gene_exp_scatter(gene)\n",
    "        print()\n",
    "    \n",
    "        print(f\"All genes info expression=0 after knocked-down:\")\n",
    "        to_zero=self.__knock_gene_d[self.__knock_gene_d['knocked']==0]\n",
    "        sns.displot(to_zero['original'])\n",
    "        plt.title('Original expression level of genes after perturbation is 0')\n",
    "        plt.xlabel('original gene counts')\n",
    "        plt.show()\n",
    "        print(f'gene number(after perturbation expression=0):{len(to_zero.index)}')\n",
    "        print('gene list(after perturbation expression=0):')\n",
    "        print(to_zero.index)\n",
    "\n",
    "analyse=SCPerturbationAnalysis('/home/rui/NFS/mount/Genome-scale_Perturb-seq/rpe1_raw_singlecell_01.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "anndata = sc.read_h5ad('/home/rui/NFS/mount/Genome-scale_Perturb-seq/rpe1_raw_singlecell_01.h5ad')\n",
    "expression = pd.DataFrame(anndata.X, columns=anndata.var.index, index=anndata.obs.index).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cell_barcode</th>\n",
       "      <th>AAACCCAAGAAACTAC-53</th>\n",
       "      <th>AAACCCAAGAAGCCAC-51</th>\n",
       "      <th>AAACCCAAGAAGCGAA-32</th>\n",
       "      <th>AAACCCAAGAATACAC-44</th>\n",
       "      <th>AAACCCAAGAATCGAT-43</th>\n",
       "      <th>AAACCCAAGACAGCGT-12</th>\n",
       "      <th>AAACCCAAGACAGCGT-52</th>\n",
       "      <th>AAACCCAAGACATCAA-1</th>\n",
       "      <th>AAACCCAAGACCAGAC-18</th>\n",
       "      <th>AAACCCAAGACCTTTG-8</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTTGTCTCCAAGA-47</th>\n",
       "      <th>TTTGTTGTCTCGAGTA-55</th>\n",
       "      <th>TTTGTTGTCTCTCCGA-20</th>\n",
       "      <th>TTTGTTGTCTGACAGT-14</th>\n",
       "      <th>TTTGTTGTCTGATGGT-14</th>\n",
       "      <th>TTTGTTGTCTGCACCT-44</th>\n",
       "      <th>TTTGTTGTCTGGGCAC-32</th>\n",
       "      <th>TTTGTTGTCTGTCCCA-37</th>\n",
       "      <th>TTTGTTGTCTTGGTCC-17</th>\n",
       "      <th>TTTGTTGTCTTTCCGG-22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187583</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187608</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188157</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000212907</th>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <td>234.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <td>45.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>31.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <td>131.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>...</td>\n",
       "      <td>73.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>85.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8749 rows × 247914 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cell_barcode     AAACCCAAGAAACTAC-53  AAACCCAAGAAGCCAC-51  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  4.0                  0.0   \n",
       "ENSG00000187583                  1.0                  0.0   \n",
       "ENSG00000188290                  5.0                  3.0   \n",
       "ENSG00000187608                  1.0                  1.0   \n",
       "ENSG00000188157                  0.0                  1.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                 16.0                  1.0   \n",
       "ENSG00000198886                234.0                 58.0   \n",
       "ENSG00000198786                 45.0                 13.0   \n",
       "ENSG00000198695                 31.0                  7.0   \n",
       "ENSG00000198727                131.0                 55.0   \n",
       "\n",
       "cell_barcode     AAACCCAAGAAGCGAA-32  AAACCCAAGAATACAC-44  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  0.0                  1.0   \n",
       "ENSG00000187583                  0.0                  1.0   \n",
       "ENSG00000188290                  2.0                  0.0   \n",
       "ENSG00000187608                  1.0                  0.0   \n",
       "ENSG00000188157                  0.0                  0.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                  2.0                  2.0   \n",
       "ENSG00000198886                 81.0                 95.0   \n",
       "ENSG00000198786                 27.0                 33.0   \n",
       "ENSG00000198695                  4.0                  7.0   \n",
       "ENSG00000198727                 72.0                116.0   \n",
       "\n",
       "cell_barcode     AAACCCAAGAATCGAT-43  AAACCCAAGACAGCGT-12  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  0.0                  1.0   \n",
       "ENSG00000187583                  0.0                  0.0   \n",
       "ENSG00000188290                  0.0                  2.0   \n",
       "ENSG00000187608                  0.0                  1.0   \n",
       "ENSG00000188157                  0.0                  1.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                  2.0                  5.0   \n",
       "ENSG00000198886                 85.0                137.0   \n",
       "ENSG00000198786                 19.0                 39.0   \n",
       "ENSG00000198695                  2.0                  2.0   \n",
       "ENSG00000198727                 79.0                176.0   \n",
       "\n",
       "cell_barcode     AAACCCAAGACAGCGT-52  AAACCCAAGACATCAA-1  AAACCCAAGACCAGAC-18  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000188976                  6.0                 1.0                  1.0   \n",
       "ENSG00000187583                  0.0                 0.0                  0.0   \n",
       "ENSG00000188290                  5.0                 0.0                  0.0   \n",
       "ENSG00000187608                  5.0                 1.0                  0.0   \n",
       "ENSG00000188157                  0.0                 0.0                  1.0   \n",
       "...                              ...                 ...                  ...   \n",
       "ENSG00000212907                  8.0                 4.0                  3.0   \n",
       "ENSG00000198886                112.0                69.0                 59.0   \n",
       "ENSG00000198786                 33.0                31.0                 13.0   \n",
       "ENSG00000198695                  4.0                 2.0                  3.0   \n",
       "ENSG00000198727                 95.0                80.0                 52.0   \n",
       "\n",
       "cell_barcode     AAACCCAAGACCTTTG-8  ...  TTTGTTGTCTCCAAGA-47  \\\n",
       "gene_id                              ...                        \n",
       "ENSG00000188976                 0.0  ...                  1.0   \n",
       "ENSG00000187583                 0.0  ...                  0.0   \n",
       "ENSG00000188290                 0.0  ...                  0.0   \n",
       "ENSG00000187608                 0.0  ...                  0.0   \n",
       "ENSG00000188157                 0.0  ...                  0.0   \n",
       "...                             ...  ...                  ...   \n",
       "ENSG00000212907                 2.0  ...                  1.0   \n",
       "ENSG00000198886                54.0  ...                 44.0   \n",
       "ENSG00000198786                20.0  ...                  5.0   \n",
       "ENSG00000198695                 7.0  ...                  1.0   \n",
       "ENSG00000198727                61.0  ...                 73.0   \n",
       "\n",
       "cell_barcode     TTTGTTGTCTCGAGTA-55  TTTGTTGTCTCTCCGA-20  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  0.0                  2.0   \n",
       "ENSG00000187583                  1.0                  0.0   \n",
       "ENSG00000188290                  1.0                  1.0   \n",
       "ENSG00000187608                  1.0                  0.0   \n",
       "ENSG00000188157                  1.0                  0.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                  3.0                  3.0   \n",
       "ENSG00000198886                 79.0                 60.0   \n",
       "ENSG00000198786                 20.0                 17.0   \n",
       "ENSG00000198695                  9.0                  0.0   \n",
       "ENSG00000198727                 90.0                 70.0   \n",
       "\n",
       "cell_barcode     TTTGTTGTCTGACAGT-14  TTTGTTGTCTGATGGT-14  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  1.0                  2.0   \n",
       "ENSG00000187583                  0.0                  1.0   \n",
       "ENSG00000188290                  0.0                  0.0   \n",
       "ENSG00000187608                  1.0                  0.0   \n",
       "ENSG00000188157                  1.0                  1.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                  3.0                  3.0   \n",
       "ENSG00000198886                 81.0                 77.0   \n",
       "ENSG00000198786                 31.0                 22.0   \n",
       "ENSG00000198695                  0.0                  2.0   \n",
       "ENSG00000198727                126.0                 76.0   \n",
       "\n",
       "cell_barcode     TTTGTTGTCTGCACCT-44  TTTGTTGTCTGGGCAC-32  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  0.0                  0.0   \n",
       "ENSG00000187583                  0.0                  0.0   \n",
       "ENSG00000188290                  1.0                  0.0   \n",
       "ENSG00000187608                  1.0                  0.0   \n",
       "ENSG00000188157                  1.0                  0.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                 10.0                  7.0   \n",
       "ENSG00000198886                188.0                 53.0   \n",
       "ENSG00000198786                 60.0                 23.0   \n",
       "ENSG00000198695                  7.0                  5.0   \n",
       "ENSG00000198727                179.0                 61.0   \n",
       "\n",
       "cell_barcode     TTTGTTGTCTGTCCCA-37  TTTGTTGTCTTGGTCC-17  TTTGTTGTCTTTCCGG-22  \n",
       "gene_id                                                                         \n",
       "ENSG00000188976                  0.0                  0.0                  4.0  \n",
       "ENSG00000187583                  0.0                  0.0                  0.0  \n",
       "ENSG00000188290                  0.0                  0.0                  0.0  \n",
       "ENSG00000187608                  0.0                  0.0                  1.0  \n",
       "ENSG00000188157                  0.0                  0.0                  0.0  \n",
       "...                              ...                  ...                  ...  \n",
       "ENSG00000212907                  1.0                  0.0                  4.0  \n",
       "ENSG00000198886                 37.0                 57.0                 87.0  \n",
       "ENSG00000198786                 14.0                 10.0                 38.0  \n",
       "ENSG00000198695                  1.0                  0.0                  6.0  \n",
       "ENSG00000198727                 57.0                 53.0                 85.0  \n",
       "\n",
       "[8749 rows x 247914 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cell_barcode</th>\n",
       "      <th>AAACCCAAGAGAGAAC-35</th>\n",
       "      <th>AAACCCAAGCGGACAT-34</th>\n",
       "      <th>AAACCCAAGTGCACCC-42</th>\n",
       "      <th>AAACCCAAGTGCCGAA-55</th>\n",
       "      <th>AAACCCACAATGCAGG-12</th>\n",
       "      <th>AAACCCAGTAGGTTTC-5</th>\n",
       "      <th>AAACCCAGTTCAAGGG-31</th>\n",
       "      <th>AAACCCATCCCAGTGG-4</th>\n",
       "      <th>AAACCCATCGACATCA-18</th>\n",
       "      <th>AAACCCATCGGTGTAT-26</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTTGAGTCACGAG-38</th>\n",
       "      <th>TTTGTTGCAAAGCGTG-16</th>\n",
       "      <th>TTTGTTGCAACTCCCT-11</th>\n",
       "      <th>TTTGTTGCACCAACAT-38</th>\n",
       "      <th>TTTGTTGCACGACAAG-1</th>\n",
       "      <th>TTTGTTGCATTCACCC-42</th>\n",
       "      <th>TTTGTTGGTGAGTGAC-55</th>\n",
       "      <th>TTTGTTGTCCATCTCG-35</th>\n",
       "      <th>TTTGTTGTCGGCTCTT-36</th>\n",
       "      <th>TTTGTTGTCTTTCCGG-22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187583</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187608</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188157</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000212907</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <td>43.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>...</td>\n",
       "      <td>55.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <td>14.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>5.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <td>51.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>85.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8749 rows × 11485 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cell_barcode     AAACCCAAGAGAGAAC-35  AAACCCAAGCGGACAT-34  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  2.0                  0.0   \n",
       "ENSG00000187583                  0.0                  1.0   \n",
       "ENSG00000188290                  0.0                  3.0   \n",
       "ENSG00000187608                  0.0                  1.0   \n",
       "ENSG00000188157                  0.0                  0.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                  2.0                  2.0   \n",
       "ENSG00000198886                 43.0                 61.0   \n",
       "ENSG00000198786                 14.0                 19.0   \n",
       "ENSG00000198695                  5.0                 12.0   \n",
       "ENSG00000198727                 51.0                 73.0   \n",
       "\n",
       "cell_barcode     AAACCCAAGTGCACCC-42  AAACCCAAGTGCCGAA-55  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  2.0                  2.0   \n",
       "ENSG00000187583                  0.0                  1.0   \n",
       "ENSG00000188290                  0.0                  1.0   \n",
       "ENSG00000187608                  0.0                  0.0   \n",
       "ENSG00000188157                  0.0                  0.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                  1.0                  6.0   \n",
       "ENSG00000198886                 65.0                 81.0   \n",
       "ENSG00000198786                 21.0                 40.0   \n",
       "ENSG00000198695                 13.0                  3.0   \n",
       "ENSG00000198727                 48.0                 73.0   \n",
       "\n",
       "cell_barcode     AAACCCACAATGCAGG-12  AAACCCAGTAGGTTTC-5  AAACCCAGTTCAAGGG-31  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000188976                  0.0                 1.0                  0.0   \n",
       "ENSG00000187583                  0.0                 0.0                  0.0   \n",
       "ENSG00000188290                  0.0                 0.0                  1.0   \n",
       "ENSG00000187608                  0.0                 0.0                  0.0   \n",
       "ENSG00000188157                  1.0                 0.0                  1.0   \n",
       "...                              ...                 ...                  ...   \n",
       "ENSG00000212907                  3.0                 8.0                  3.0   \n",
       "ENSG00000198886                 46.0                83.0                 85.0   \n",
       "ENSG00000198786                 18.0                29.0                 31.0   \n",
       "ENSG00000198695                  1.0                 6.0                  1.0   \n",
       "ENSG00000198727                 24.0               103.0                 49.0   \n",
       "\n",
       "cell_barcode     AAACCCATCCCAGTGG-4  AAACCCATCGACATCA-18  AAACCCATCGGTGTAT-26  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000188976                 0.0                  0.0                  1.0   \n",
       "ENSG00000187583                 0.0                  0.0                  0.0   \n",
       "ENSG00000188290                 1.0                  0.0                  0.0   \n",
       "ENSG00000187608                 0.0                  0.0                  1.0   \n",
       "ENSG00000188157                 1.0                  0.0                  0.0   \n",
       "...                             ...                  ...                  ...   \n",
       "ENSG00000212907                 5.0                  0.0                  3.0   \n",
       "ENSG00000198886                75.0                 35.0                 44.0   \n",
       "ENSG00000198786                28.0                  6.0                 24.0   \n",
       "ENSG00000198695                 3.0                  1.0                  4.0   \n",
       "ENSG00000198727                93.0                 33.0                 68.0   \n",
       "\n",
       "cell_barcode     ...  TTTGTTGAGTCACGAG-38  TTTGTTGCAAAGCGTG-16  \\\n",
       "gene_id          ...                                             \n",
       "ENSG00000188976  ...                  0.0                  1.0   \n",
       "ENSG00000187583  ...                  0.0                  0.0   \n",
       "ENSG00000188290  ...                  1.0                  0.0   \n",
       "ENSG00000187608  ...                  0.0                  2.0   \n",
       "ENSG00000188157  ...                  0.0                  1.0   \n",
       "...              ...                  ...                  ...   \n",
       "ENSG00000212907  ...                  5.0                  2.0   \n",
       "ENSG00000198886  ...                 55.0                 95.0   \n",
       "ENSG00000198786  ...                 23.0                 26.0   \n",
       "ENSG00000198695  ...                  0.0                  0.0   \n",
       "ENSG00000198727  ...                 59.0                125.0   \n",
       "\n",
       "cell_barcode     TTTGTTGCAACTCCCT-11  TTTGTTGCACCAACAT-38  TTTGTTGCACGACAAG-1  \\\n",
       "gene_id                                                                         \n",
       "ENSG00000188976                  1.0                  1.0                 2.0   \n",
       "ENSG00000187583                  0.0                  0.0                 0.0   \n",
       "ENSG00000188290                  0.0                  0.0                 0.0   \n",
       "ENSG00000187608                  1.0                  0.0                 0.0   \n",
       "ENSG00000188157                  0.0                  0.0                 1.0   \n",
       "...                              ...                  ...                 ...   \n",
       "ENSG00000212907                  4.0                  3.0                 2.0   \n",
       "ENSG00000198886                 73.0                 63.0                46.0   \n",
       "ENSG00000198786                 26.0                 29.0                21.0   \n",
       "ENSG00000198695                  7.0                  4.0                 3.0   \n",
       "ENSG00000198727                 64.0                 72.0                61.0   \n",
       "\n",
       "cell_barcode     TTTGTTGCATTCACCC-42  TTTGTTGGTGAGTGAC-55  \\\n",
       "gene_id                                                     \n",
       "ENSG00000188976                  1.0                  0.0   \n",
       "ENSG00000187583                  0.0                  1.0   \n",
       "ENSG00000188290                  3.0                  3.0   \n",
       "ENSG00000187608                  0.0                  0.0   \n",
       "ENSG00000188157                  0.0                  0.0   \n",
       "...                              ...                  ...   \n",
       "ENSG00000212907                  4.0                  6.0   \n",
       "ENSG00000198886                 57.0                113.0   \n",
       "ENSG00000198786                 22.0                 26.0   \n",
       "ENSG00000198695                  4.0                  3.0   \n",
       "ENSG00000198727                 54.0                 94.0   \n",
       "\n",
       "cell_barcode     TTTGTTGTCCATCTCG-35  TTTGTTGTCGGCTCTT-36  TTTGTTGTCTTTCCGG-22  \n",
       "gene_id                                                                         \n",
       "ENSG00000188976                  0.0                  1.0                  4.0  \n",
       "ENSG00000187583                  0.0                  0.0                  0.0  \n",
       "ENSG00000188290                  0.0                  0.0                  0.0  \n",
       "ENSG00000187608                  0.0                  0.0                  1.0  \n",
       "ENSG00000188157                  0.0                  0.0                  0.0  \n",
       "...                              ...                  ...                  ...  \n",
       "ENSG00000212907                  0.0                  3.0                  4.0  \n",
       "ENSG00000198886                  0.0                 75.0                 87.0  \n",
       "ENSG00000198786                  5.0                 15.0                 38.0  \n",
       "ENSG00000198695                  3.0                  1.0                  6.0  \n",
       "ENSG00000198727                  0.0                 71.0                 85.0  \n",
       "\n",
       "[8749 rows x 11485 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression[anndata.obs[anndata.obs['gene'] == 'non-targeting'].index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting dill\n",
      "  Downloading dill-0.3.7-py3-none-any.whl (115 kB)\n",
      "\u001b[K     |████████████████████████████████| 115 kB 201 kB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: dill\n",
      "Successfully installed dill-0.3.7\n"
     ]
    }
   ],
   "source": [
    "! pip install dill\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### scatter and vocano plot for each perturbation\n",
    "total_mRNA=analyse.total_mRNA_info\n",
    "expression=analyse._SCPerturbationAnalysis__expression\n",
    "anndata=analyse._SCPerturbationAnalysis__anndata\n",
    "downstream_gene_list=downstream_gene_list\n",
    "\n",
    "def scatter_vocano_for_gene(gene,output_expression_is_0):\n",
    "    expriment=expression[anndata.obs[anndata.obs['gene']==gene].index]\n",
    "    control=expression[anndata.obs[anndata.obs['gene']=='non-targeting'].index].mean(axis=1)\n",
    "    control.index=analyse.gene_lfc.columns\n",
    "    %matplotlib inline\n",
    "    plt.hist(control)\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "    %matplotlib\n",
    "    # control=self.__self.__expression[self.__self.__anndata.obs[self.__self.__anndata.obs['gene']=='non-targeting'].index].sample(n=expriment.shape[1],axis=1).mean(axis=1)\n",
    "    expriment1=expriment.mean(axis=1)\n",
    "    expriment1.index=analyse.gene_lfc.columns\n",
    "    print(f\"knocked gene\", gene)\n",
    "    print(f\"cell number:\",expriment.shape[1])\n",
    "\n",
    "    ### scatter plot and valcano\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12,6))\n",
    "    plt.rcParams.update({'font.size': 10})\n",
    "    sct=ax1.scatter(np.log2(control),(np.log2(expriment1)),s=10)\n",
    "    downstream_genes=downstream_gene_list[gene]\n",
    "    ax1.scatter(np.log2(control[control.index.isin(downstream_genes)]),(np.log2(expriment1[control.index.isin(downstream_genes)])),edgecolors='orange',s=10, label='downstream genes')\n",
    "    ax1.scatter(np.log2(control[gene]),(np.log2(expriment1[gene])),color='red',s=10, label='downstream genes')\n",
    "    cursor = mplcursors.cursor(sct, hover=True)\n",
    "    # by default the annotation displays the xy positions\n",
    "    @cursor.connect(\"add\")\n",
    "    def on_add(sel):\n",
    "        sel.annotation.set_text([control.index[sel.index],gene_exp.loc[gene,control.index[sel.index]]\n",
    "                                ,gene_exp.loc['non-targeting',control.index[sel.index]]])\n",
    "\n",
    "    ax1.plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control)),np.log2(max(control))],color='red',linestyle='--',label='y=x')\n",
    "    ratio=total_mRNA.loc[gene,'ratio']\n",
    "    ax1.plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control))*ratio,np.log2(max(control))*ratio],color='red',label=f\"total_mRNS_ratio={ratio:.2f}\")\n",
    "    ax1.set_xlabel('log2(original UMI count)')\n",
    "    ax1.set_ylabel('log2(perturbed UMI count)')\n",
    "    ax1.set_title('original VS knocked-down gene expression')\n",
    "    ax1.legend(loc='upper left',bbox_to_anchor=(0,1), frameon=False)\n",
    "    # return 敲完表达为0的基因，敲完表达变化大的基因\n",
    "\n",
    "\n",
    "    p_values=analyse.gene_pvalue.loc[gene]\n",
    "    log2_fold_change = analyse.gene_lfc.loc[gene]\n",
    "    data = pd.DataFrame({'log2_fold_change': log2_fold_change, 'p_value': p_values},index=analyse.gene_lfc.columns)\n",
    "    p_value_threshold = 0.05\n",
    "    fold_change_threshold = 1.5\n",
    "    voc=ax2.scatter(data['log2_fold_change'], -np.log10(data['p_value']),color='gray', alpha=0.7)\n",
    "    plt.scatter(data[data.index.isin(downstream_genes)]['log2_fold_change'], -np.log10(data[data.index.isin(downstream_genes)]['p_value']), color='red', alpha=0.7)\n",
    "    cursor = mplcursors.cursor(voc, hover=True)\n",
    "    @cursor.connect(\"add\")\n",
    "    def on_add(sel):\n",
    "        sel.annotation.set_text(data.index[sel.index])\n",
    "    ax2.axhline(-np.log10(p_value_threshold), color='blue', linestyle='--', label='p-value threshold')\n",
    "    ax2.axvline(fold_change_threshold, color='green', linestyle='--', label='fold change threshold')\n",
    "    ax2.axvline(-fold_change_threshold, color='green', linestyle='--')\n",
    "    ax2.set_xlabel('log2 Fold Change')\n",
    "    ax2.set_ylabel('-log10(p-value)')\n",
    "    ax2.set_title('Gene Volcano Plot')\n",
    "    if output_expression_is_0==True:\n",
    "        expr=pd.DataFrame(expriment1,columns=['UMI'])\n",
    "        zero_gene=expr[expr['UMI']==0].index\n",
    "        print(f'genes with 0 expression after perturbation)')\n",
    "        return zero_gene\n",
    "\n",
    "scatter_vocano_for_gene('IMP4',True)\n",
    "# CBX1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No handles with labels found to put in legend.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: Qt5Agg\n"
     ]
    }
   ],
   "source": [
    "### global scatter plot\n",
    "total_mRNA=analyse.total_mRNA_info\n",
    "knock_gene_df=analyse.knocked_gene_info\n",
    "total_mRNA_hue=total_mRNA[total_mRNA.index.isin(knock_gene_df.index)]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "%matplotlib\n",
    "\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12,6), gridspec_kw={'width_ratios': [3, 2]})\n",
    "ax1.tick_params(axis='both',)\n",
    "ax1.set_title('original VS knocked-down gene expression level comparison')\n",
    "plt.rcParams.update({'font.size': 10})\n",
    "# scatter plot\n",
    "sc = ax1.scatter(np.log2(knock_gene_df['original']),np.log2(knock_gene_df['knocked']),c=total_mRNA_hue['ratio'].rename('total mRNA ratio'),cmap='bwr')\n",
    "ax1.plot([-9,max(np.log2(knock_gene_df['original']))],[-9,max(np.log2(knock_gene_df['original']))],color='gray',linestyle='--')\n",
    "# cax=plt.axes([0.95,0.1,0.03,0,8])\n",
    "cbar=plt.colorbar(sc)\n",
    "# cbar.set_lable('Color value')\n",
    "# axis 1 labels\n",
    "ax1.set_xlabel('log2(original gene expression)')\n",
    "ax1.set_ylabel('log2(knocked-down gene experssion)')\n",
    "ax2.set_xticks([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_title('Detail info')\n",
    "ax2.set_xlim(0,2)\n",
    "\n",
    "cursor = mplcursors.cursor(sc, hover=True)# by default the annotation displays the xy positions\n",
    "@cursor.connect(\"add\")\n",
    "def on_add(sel):\n",
    "    sel.annotation.set_text(knock_gene_df.index[sel.index])\n",
    "    for text in ax2.texts:\n",
    "        text.set_visible(False)\n",
    "    source =knock_gene_df.index[sel.index]\n",
    "    # ax2.text(0.03, 0.2, source, ha='left', va='top', fontsize=7)\n",
    "    ax2.axhline(0.3,lw=0.2,color='black',linestyle='--')\n",
    "    ax2.text(0.03,0.9,\"original expresison:\")\n",
    "    ax2.text(2,0.9,analyse._SCPerturbationAnalysis__gene_exp.loc['non-targeting',knock_gene_df.index[sel.index]],ha='right')\n",
    "    ax2.text(0.03,0.8,\"knocked expresison:\")\n",
    "    ax2.text(2,0.8,analyse._SCPerturbationAnalysis__gene_exp.loc[knock_gene_df.index[sel.index],knock_gene_df.index[sel.index]],ha='right')\n",
    "    ax2.text(0.03,0.7,\"original total mRNA:\")\n",
    "    ax2.text(2,0.67,total_mRNA.loc['non-targeting','total_mRNA'],ha='right')\n",
    "    ax2.text(0.03,0.6,\"knocked total mRNA:\")\n",
    "    ax2.text(2,0.57,total_mRNA.loc[knock_gene_df.index[sel.index],'total_mRNA'],ha='right')  \n",
    "    ax2.text(0.03,0.5,\"amount of downstream genes:\")\n",
    "    ax2.text(2,0.5,analyse._SCPerturbationAnalysis__downstream_counts.loc[knock_gene_df.index[sel.index],'counts'],ha='right')\n",
    "    ax2.text(0.03,0.4,\"amount of upstream genes:\")\n",
    "    ax2.text(2,0.4,upstream_num.iloc[sel.index,0],ha='right')\n",
    "    ax2.text(1,0.25,\"Colors in scatterplot: total mRNA ratio\",ha='center')\n",
    "    ax2.text(1,0.15,\"For more detailed infomation,\",ha='center')\n",
    "    ax2.text(1,0.12,\"please run:\",ha='center')\n",
    "    ax2.text(1,0.09,\"_.get_all_info_for_perturbation(gene)\",ha='center')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 寻找第一类TF，返回其下游基因数量和list\n",
    "\n",
    "# Pre-calculate the threshold for all rows\n",
    "non_targeting = gene_exp.loc['non-targeting']\n",
    "thresholds = pd.DataFrame(np.percentile(np.abs(gene_exp - non_targeting), 80, axis=1),index=lfc.index)\n",
    "\n",
    "downstream_counts = []\n",
    "genelist = []\n",
    "num = 0\n",
    "\n",
    "for i in lfc.index:\n",
    "    abs_diff_non_targeting = np.abs(non_targeting - gene_exp.loc[i])\n",
    "    data=np.full((len(abs_diff_non_targeting),1),thresholds.loc[i])\n",
    "    diff=pd.DataFrame(data,index=abs_diff_non_targeting.index)\n",
    "    #\n",
    "    mask = ((np.abs(lfc.loc[i]) > 1.5).values)&(abs_diff_non_targeting > diff.iloc[:,0]) & ((p_value.loc[i].values < 0.05))\n",
    "    gene_list = lfc.columns[mask].tolist()\n",
    "    genelist.append(gene_list)\n",
    "downstream_gene_list={key: values for key,values in zip(lfc.index, genelist)}\n",
    "downstream_gene_list={key: [item for item in value if item!=key] for key,value in downstream_gene_list.items()}\n",
    "downstream_num=[]\n",
    "for key,value in downstream_gene_list.items():\n",
    "    downstream_num.append(len(value))\n",
    "downstream_gene_counts=pd.DataFrame(downstream_num,index=lfc.index,columns=['counts'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8749"
      ]
     },
     "execution_count": 370,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(downstream_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GNPNAT1']"
      ]
     },
     "execution_count": 367,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "downstream_gene_list['CBX1']\n",
    "# ['CBX1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 查找上游基因\n",
    "def find_keys_containing_elements(data_dict, specific_elements):\n",
    "    result_dict = {element: [] for element in specific_elements}\n",
    "    for key in data_dict:\n",
    "        for element in specific_elements:\n",
    "            if element in data_dict[key]:\n",
    "                result_dict[element].append(key)\n",
    "    return result_dict\n",
    "observe_gene=lfc.columns\n",
    "dict=find_keys_containing_elements(downstream_gene_list,observe_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8749"
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upstream_num=[]\n",
    "for key,value in dict.items():\n",
    "    upstream_num.append(len(value))\n",
    "len(upstream_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [],
   "source": [
    "### knocked gene expression difference distribution and volcano plot\n",
    "# diff_exp=analyse.knocked_gene_info['variation']\n",
    "# sns.scatterplot(analyse.knocked_gene_info['original'],analyse.knocked_gene_info['knocked'])\n",
    "total_mRNA_hue=total_mRNA[total_mRNA.index.isin(analyse.knocked_gene_info.index)]\n",
    "sns.scatterplot(np.log2(analyse.knocked_gene_info['original']),np.log2(analyse.knocked_gene_info['knocked']),hue=total_mRNA_hue['ratio'].rename('total mRNA ratio'),palette='Oranges')\n",
    "abnormal=analyse.knocked_gene_info[analyse.knocked_gene_info['variation']>0]\n",
    "# sns.scatterplot(np.log2(abnormal['original']),np.log2(abnormal['knocked']),color='red',label='abnormal')\n",
    "plt.plot([-9,max(np.log2(analyse.knocked_gene_info['original']))],[-9,max(np.log2(analyse.knocked_gene_info['original']))],color='gray',linestyle='--')\n",
    "plt.xlim(-9,8)\n",
    "plt.xlabel('log2(original expression)')\n",
    "plt.ylabel('log2(knocked experssion)')\n",
    "plt.title('knocked gene experssion level')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anndata=analyse._SCPerturbationAnalysis__anndata\n",
    "expression=analyse._SCPerturbationAnalysis__expression\n",
    "total_mRNA=analyse.total_mRNA_info\n",
    "gene_lfc=analyse.gene_lfc\n",
    "gene_p_value=analyse.gene_pvalue\n",
    "gene_p_value.columns=gene_lfc.columns\n",
    "\n",
    "def observe_gene_exp_scatter(gene):\n",
    "    expriment=expression[anndata.obs[anndata.obs['gene']==gene].index]\n",
    "    control=expression[anndata.obs[anndata.obs['gene']=='non-targeting'].index].mean(axis=1)\n",
    "    # control=self.__expression[self.__anndata.obs[self.__anndata.obs['gene']=='non-targeting'].index].sample(n=expriment.shape[1],axis=1).mean(axis=1)\n",
    "    expriment1=expriment.mean(axis=1)\n",
    "    print(f\"knocked gene\", gene)\n",
    "    print(f\"cell number:\",expriment.shape[1])\n",
    "    \n",
    "    ### scatter plot and valcano\n",
    "    fig, axes = plt.subplots(1, 2, figsize=(8,4))\n",
    "    sns.scatterplot(np.log2(control),(np.log2(expriment1)),ax=axes[0])\n",
    "    axes[0].plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control)),np.log2(max(control))],color='red',linestyle='--',label='y=x')\n",
    "    ratio=total_mRNA.loc[gene,'ratio']\n",
    "    axes[0].plot([np.log2(min(control)),np.log2(max(control))],[np.log2(min(control))*ratio,np.log2(max(control))*ratio],color='red',label=f\"total_mRNS_ratio={ratio:.2f}\")\n",
    "    axes[0].set_xlabel('log2(original gene count)')\n",
    "    axes[0].set_ylabel('log2(perturbed gene count)')\n",
    "    axes[0].legend()\n",
    "\n",
    "\n",
    "    p_values=gene_p_value.loc[gene]\n",
    "    log2_fold_change = gene_lfc.loc[gene]\n",
    "    data = pd.DataFrame({'log2_fold_change': log2_fold_change, 'p_value': p_values})\n",
    "    p_value_threshold = 0.05\n",
    "    fold_change_threshold = 1.5\n",
    "    sns.scatterplot(data['log2_fold_change'], -np.log10(data['p_value']), c=np.where((data['p_value'] < p_value_threshold) & (np.abs(data['log2_fold_change']) > fold_change_threshold), 'red', 'gray'), alpha=0.7,ax=axes[1])\n",
    "    axes[1].axhline(-np.log10(p_value_threshold), color='blue', linestyle='--', label='p-value threshold')\n",
    "    axes[1].axvline(fold_change_threshold, color='green', linestyle='--', label='fold change threshold')\n",
    "    axes[1].axvline(-fold_change_threshold, color='green', linestyle='--')\n",
    "    axes[1].set_xlabel('log2 Fold Change')\n",
    "    axes[1].set_ylabel('-log10(p-value)')\n",
    "    axes[1].set_title('Gene Volcano Plot')\n",
    "    plt.show()\n",
    "    return data\n",
    "\n",
    "\n",
    "### look into abnormal perturbations  'NUMA1'\n",
    "# observe_gene_exp_scatter('NUMA1')\n",
    "abnormal=analyse.knocked_gene_info[analyse.knocked_gene_info['variation']>0]\n",
    "for gene in abnormal.index:\n",
    "    observe_gene_exp_scatter(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "analyse.knocked_gene_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 6.79999999999999, 'original gene counts')"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXMAAAFwCAYAAABKLps8AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAgNUlEQVR4nO3deZhkZXn38e89DDDsizPgADPggguY1w0VESMJqANqQAWFKELE4JL4SoxGYhITjRr09Y2YRQWNYRBFUEEQCUoQJMrmiIgQMBgyCGEblB1FB+788Twth04v1c2crp6nv5/rqqurznqf7VdPnTp1OjITSdLabd6wC5AkPXKGuSQ1wDCXpAYY5pLUAMNckhpgmEtSA2ZtmEfEuyPi02t62AGmlRHx+DUxrdkkIu6JiMf2MN2VEbHXmp7uqHkcFxHvn+a4L4+I6+vyP31N1zZsEfHmiLilLt+jhl3PTFnT+12Px8eVEbHHmp7uWGYkzCPi0Ij4YUTcFxE3R8QnImLzicbJzA9m5hsGmf5Uhp2rMnPjzLx22HUMwUeAP6zL//1hF7MmRcS6wN8CL8rMjYHfiIgbhlzWhCJij2HXGBHnRcTD8qKv4yMzd87M86Y6XkTsGRFX18w8NyK2n2yc3sM8Iv4Y+BDwTmAzYFdge+DsiFhvnHHm913XbDWXl70n2wNXDruInmwNLGANLV/f+96amP5cOD4iYiFwCvAXwJbACuCkSUfMzN4ewKbAPcCrRnXfGLgVeH19/VfAl4ATgLuAN9RuJ3TGeR1wHfDTupArgb06459Qn+8AJHAI8BPgNuDPOtN5NnAhcAdwE/APwHqd/gk8fpzl2Qz4pzrefwPvB9YB1gMuA95ah1sH+A7wnlHLdxJwN3Ap8NTOdFcC7wIuB+4H5lPe9C6odf4A2KMz/KHAtXVa/wW8pnZ/PPAt4M663CeNtVx1OY4HVtV1+ufAvM60v01p0d5ep7/3BNu4ux3mAUcC/1m308nAlrXfWZQWcnfcHwCvqM+fBJwN/Az4EZ19BjgOeP84859X67+Osk8dX5dvfcq+l8C9wH+OM/6L6vzuBD5e198bOv1fD1xV18XXge1HrdM3AdfU/v8IxGTjAgF8tNZ7Z93uTxmnvt+r07i7bvM31u5PqMuVdTnPBX4OPFhf3wNsM8k22aGOfxjlWDl/jPnvAdwAvLvuUyup+1vtv37dV34C3AJ8Ethg1LjvAm4GvjhOjQ/bviPjTXJ8rAT+FPj3un7/GVhQh98COIOyf99en29X+30AeAD4RZ3/P8zg8fFsSjDfVdfV344zzuHABZ3XG9X19qQJ87avIK9FLANWA/PH6LccOLETdr8C9qs73wY8PKB3qit+d0pwfqQOP1GYf6pO56l1B3hy7f9MSlDOr8NeBRwxYJh/BTimrtytgEt46OB6St24Twb+DLgIWGfU8u0PrAu8o+4E63Y2+GXAklrztpQDb5+6Pl5YXy+q874LeGIddzGwc31+Yp33PEqLbfdxwvx44DRgk7oO/gM4rLOz/gr4fcqb0puBG+mE1AQ76xF1ubejHOTHdLbx64DvdMbbifJGtX5dpuspwTUfeAYlOEaW6zjGD/PXAz8GHktpJJwCfHbA7bmwrstX1Pm+rS77G2r//eq0n1z7/zkPP8iSEhSbA0spB/+yycYFXgx8r44XdZjF49T4EuBxdbgXAPcBzxi1r88fKwQH2CYj4x9ft8EG44T5asrpnPVrDffy0P53NHA6pQW5CfBV4G9GjfuhOu4G49T4sO07ehhGHR+dblfUbltSGk/vr/0eBbwS2LDW9EXgK53pnUfnDXsGj48LgYPr842BXccZ52PAJ0Z1uwJ45YR520eIdwp4LXDzOP2OAs7uhN35o/r/FQ8F9HtGdsD6ekPgl0wc5tt1hr8EOHCcOo4ATp3s4Kd8pL2fzg4PHASc23n9x8DVlFDfcdSyXNR5PY/Sun9+Z4O/vtP/XXQCqXb7OuXTxkaUEHwlow6+uhMe21320ctVd8D7gZ06/d4InNfZWX88al0n8OgBdtargD07/RbXHX8+5cC4l4dapx8APlOfvxr4t1HTPQb4yxzjYB813DnAWzqvnzgyz4m2Z+33OuDCzuugvKmMhPm/UA/izna7r7MMycPfME8GjpxsXOC3KQGxK7XFN4Vj6ivA20bt6xOF+UTbZGT8x04wvz0ogbzRqOX8i7q+7gUe1+n3XOC/OuP+ktpinqDGh23f0cMw6vjodHtT5/U+jP/p62nA7Z3X5zFOmNPv8XE+8F5g4STb+J+Ao0Z1+w5w6ETj9X3O/DZg4TjnuRbX/iOun2A623T7Z+Z9lJbqRG7uPL+P8k5IRDwhIs6oX8TeBXyQ0kKbzPaUVvVNEXFHRNxBCZytOsMspxwgZ2bmNaPG79b/IOXj5zZj9a/zOmBkPnVeu1Nab/dSwu9NtZavRcST6nh/QjnALqnfor9+jOVYSPl0c12n23WUTwMjfr3u6rqGuv4msT1waqfmqygfabfOzLuBrwEH1mEPBD7XGe85o5b3NcCjB5jnNmMsy3zKm+8g43a3S1K2S3d5Ptap6WeU9TvmuqKzn000bmZ+k3J67x+BWyLi2IjYdKwCI2LviLgoIn5Wp7MPg+2v3WUYc5t0hpno2IMShPd2Xl9HWXeLKGH2vc70z6rdR6zKzF9Mod7xjFVjt9tITUTEhhFxTERcV4/x84HNI2KdAebT5/FxGOX02NUR8d2IeOk4w91DOUXdtSnlVNu4+g7zCynvcq/odoyIjYC9Ka2qETnBdG6ifEwcGX8Dykep6fgEpfW8Y2ZuSjkXGAOMdz1lWRZm5ub1sWlm7twZ5uOUj90vjojdR42/pFP/PMry3Njp313+6ykt8807j40y8yiAzPx6Zr6Q8oZ4NeWUEpl5c2b+fmZuQ2lNfHyMyyxvo7TMtu90W0r5DuCRup5y/rBb94LMHJn2icBBEfFcykfuczvjfWvUeBtn5psHmOeNYyzLaso5ycmM3q+i+7rW9cZRdW2QmRcMMO0Jx83Mv8vMZwI7Uw7wd46eQESsD3yZclpx68zcHDiT8ffXsY6hybbJeON1bVGP2RFLKev9Nsq53J07094sy5U14017rHndS3lTGDHWm/hY4y3pPB+pCcon5CcCz6nH+G/W7iPrbaLl7e34yMxrMvMgSgPwQ8CXRq3XEVdSTg8Dv87LxzHJF929hnlm3kn5WPH3EbEsItaNiB0o57BuAD474KS+BLwsInarV8C8l8ECeCybUM6T3lNbtIMEBpl5E/AN4P9HxKYRMS8iHhcRLwCIiIMp5+MPBf4vsDwiujv1MyPiFfVTyhGUN4aLxpndCZTlfXFErBMRC+olXdtFxNYR8Tt1A99PeRd/oNZwQESMhNHtlJ32gVHL8QDlY/IHImKTesnT2+s8H6lP1uluX+tZFBH7dvqfSTlI3kf5cvbB2v0M4AkRcXDdR9aNiGdFxJMHmOeJwB9FxGPq+v5gnfbqAcb9GuVyvv3qdvkDHh4knwT+NCJ2rsuzWUQcMMB0Jxy3Lttz6qWF91K+jHtgjGmsRznXvApYHRF7U76wHc8twKMiYrNRdUy0TQb13ohYLyKeD7wU+GLdfp8CPhoRW9XpbxsRL55ijZcB+0TElhHxaMrxMYg/qMfElpRG2cgVH5tQ3mTuqP3+cowaxrymvM/jIyJeGxGL6nq7o3Yea7ufCjwlIl4ZEQsop5kvz8yrJ5p+75cmZuaHKSv6I5QQvZjSWtgzM+8fcBpXAm8FvkBpTd1NuRJgoPFHeQfwu3Uan2KQS34e8jrKATbyDfqXgMURsZTyRdDrMvOezPw85Vvrj3bGPY1yeuR24GDKVRy/GmsmmXk9sC9lva2irK93UrbXPErL40bKR/cXAG+poz4LuDgi7qF8KfW2zPyvMWbxVkqIXEv5Zv7zwGemsB7G87E6329ExN2UN6vndJbrfsoXlHvVeY50v5sSUgfW5bqZh740m8xnKI2C8ylfKv+CsnyTyszbgAOAD1NO2+1E2W731/6n1jq+UD+uX0H5RDnItCcad1PKvnc7D12h9ZExpnE3pWFwch32dynrd7x5Xk15c7u2nvbYhkm2yYBurvO/kXJq7E2dYHkX5Yvei+py/iulVTyVGj9LubJpJaXBNOgx+fk6/LX1MfLDsqMpn/xuoyzvWaPG+xiwf0TcHhF/N8Z0+zo+lgFX1uPzY5Tv8f7XKajMXEX5TuwDlPX+HB46PTmuqCfX1yq1BXYH5VTJWGE1q0TEX1G+hHvtsGvR+Orprxsol96dO9nwc0GUXy+ekJnbTTKohmzW/px/tIh4Wf1iYyNKK+aHlHdyadrqqazN6/npke9Pxjv9Jc1aa02YU0473FgfO1I+oqx9Hys02zyX8oOa24CXAftl5s+HW5I0dWvlaRZJ0sOtTS1zSdI41oqb1ixbtizPOmv0F9KStNaY7qXUA1srWua33Xbb5ANJ0hy2VoS5JGlihrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMJekBhjmktQAw1ySGtB0mG+7ZCkRMdBj2yVLh12uJE3bWnE/8+m68YbrefUxFww07Elv3K3naiSpP023zCVprjDMJakBhrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMB8xb7435ZK01mr6RltT8uBqb8olaa1ly1ySGmCYS1IDDHNJaoBhLkkNMMwlqQGGuSQ1wDCXpAYY5pLUAMNckhpgmEtSAwxzSWqAYS5JDTDMJakBhrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ3oPcwjYp2I+H5EnFFfbxkRZ0fENfXvFn3XIEmtm4mW+duAqzqvjwTOycwdgXPqa0nSI9BrmEfEdsBLgE93Ou8LLK/PlwP79VmDJM0FfbfMjwb+BHiw023rzLwJoP7daqwRI+LwiFgREStWrVrVc5mStHbrLcwj4qXArZn5vemMn5nHZuYumbnLokWL1nB1ktSW+T1O+3nA70TEPsACYNOIOAG4JSIWZ+ZNEbEYuLXHGiRpTuitZZ6Zf5qZ22XmDsCBwDcz87XA6cAhdbBDgNP6qkGS5ophXGd+FPDCiLgGeGF9LUl6BPo8zfJrmXkecF59/lNgz5mYryTNFf4CVJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMJekBhjmktQAw1ySGmCYS1IDDHNJaoBhLkkNMMwlqQGGuSQ1wDCXpAYY5pLUAMNckhpgmEtSAwxzSWqAYS5JDTDMJakBhrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMJekBhjmktQAw1ySGmCYS1IDDHNJaoBhLkkNMMwlqQGGuSQ1wDCXpAYY5pLUAMNckhpgmEtSAwxzSWqAYS5JDTDMJakBhrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMJekBvQW5hGxICIuiYgfRMSVEfHe2n3LiDg7Iq6pf7foqwZJmiv6bJnfD/x2Zj4VeBqwLCJ2BY4EzsnMHYFz6mtJ0iPQW5hncU99uW59JLAvsLx2Xw7s11cNkjRX9HrOPCLWiYjLgFuBszPzYmDrzLwJoP7dapxxD4+IFRGxYtWqVX2WKUlrvV7DPDMfyMynAdsBz46Ip0xh3GMzc5fM3GXRokW91ShJLZiRq1ky8w7gPGAZcEtELAaof2+diRokqWV9Xs2yKCI2r883APYCrgZOBw6pgx0CnNZXDZI0V8zvcdqLgeURsQ7lTePkzDwjIi4ETo6Iw4CfAAf0WIMkzQm9hXlmXg48fYzuPwX27Gu+kjQX+QtQSWqAYS5JDTDMJakBhrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMJekBhjmPdt2yVIiYqDHtkuWDrtcSWupPu9nLuDGG67n1cdcMNCwJ71xt56rkdQqW+aS1ADDXJIaYJhLUgMMc0lqwEBhHhHPG6TbnDFv/sBXqEjSTBj0apa/B54xQLe54cHVXqEiaVaZMMwj4rnAbsCiiHh7p9emwDp9FiZJGtxkLfP1gI3rcJt0ut8F7N9XUZKkqZkwzDPzW8C3IuK4zLxuhmqSJE3RoOfM14+IY4EduuNk5m/3UZQkaWoGDfMvAp8EPg080F85kqTpGDTMV2fmJ3qtRJI0bYP+aOirEfGWiFgcEVuOPHqtTJI0sEFb5ofUv+/sdEvgsWu2HEnSdAwU5pn5mL4LkSRN30BhHhGvG6t7Zh6/ZsuRJE3HoKdZntV5vgDYE7gUMMwlaRYY9DTLW7uvI2Iz4LO9VCRJmrLp3gL3PmDHNVmIJGn6Bj1n/lXK1StQbrD1ZODkvoqSJE3NoOfMP9J5vhq4LjNv6KEeSdI0DHSapd5w62rKnRO3AH7ZZ1GSpKkZ9D8NvQq4BDgAeBVwcUR4C1xJmiUGPc3yZ8CzMvNWgIhYBPwr8KW+CpMkDW7Qq1nmjQR59dMpjCtJ6tmgLfOzIuLrwIn19auBM/spSZI0VZP9D9DHA1tn5jsj4hXA7kAAFwKfm4H6JEkDmOxUydHA3QCZeUpmvj0z/4jSKj+639IkSYOaLMx3yMzLR3fMzBWUfyEnSZoFJgvzBRP022BNFiJJmr7Jwvy7EfH7oztGxGHA9/opSZI0VZNdzXIEcGpEvIaHwnsXYD3g5T3WJUmaggnDPDNvAXaLiN8CnlI7fy0zv9l7ZZKkgQ16P/NzgXN7rkWSNE3+ilOSGmCYS1IDDHNJaoBhLkkNMMwlqQG9hXlELImIcyPiqoi4MiLeVrtvGRFnR8Q19e8WfdUgSXNFny3z1cAfZ+aTgV2BP4iInYAjgXMyc0fgnPpakvQI9BbmmXlTZl5an98NXAVsC+wLLK+DLQf266sGSZorZuSceUTsADwduJhyf/SboAQ+sNVM1CBJLes9zCNiY+DLwBGZedcUxjs8IlZExIpVq1b1V6AkNaDXMI+IdSlB/rnMPKV2viUiFtf+i4Fbxxo3M4/NzF0yc5dFixb1WaYkrfX6vJolgH8CrsrMv+30Oh04pD4/BDitrxokaa4Y9B86T8fzgIOBH0bEZbXbu4GjgJPrPdF/AhzQYw2SNCf0FuaZ+W3KP38ey559zVeS5iJ/ASpJDTDMJakBhrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMJekBhjmktQAw1ySGmCYS1IDDHNJaoBhLkkNMMwlqQGGuSQ1wDCXpAYY5pLUAMNckhpgmEtSAwxzSWqAYS5JDTDMJakBhrkkNcAwl6QGGOaS1ADDXJIaYJhLUgMMc0lqgGEuSQ0wzCWpAYa5JDXAMJekBhjmktQAw1ySGmCYS1IDDHNJaoBhLkkNMMwlqQGGuSQ1wDCXpAYY5pLUAMNckhpgmEtSAwxzSWqAYS5JDTDMJakBhrkkNcAwl6QGGOZzwLZLlhIRAz22XbJ02OVKmob5wy5A/bvxhut59TEXDDTsSW/credqJPXBlrkkNaC3MI+Iz0TErRFxRafblhFxdkRcU/9u0df8JWku6bNlfhywbFS3I4FzMnNH4Jz6WpL0CPUW5pl5PvCzUZ33BZbX58uB/fqavyTNJTN9znzrzLwJoP7darwBI+LwiFgREStWrVo1YwVK0tpo1n4BmpnHZuYumbnLokWLhl2OJM1qMx3mt0TEYoD699YZnr8kNWmmw/x04JD6/BDgtBmevyQ1qc9LE08ELgSeGBE3RMRhwFHACyPiGuCF9bUk6RHq7RegmXnQOL327GuekjRXzdovQCVJgzPM11JTuXmWpPZ5o621lDfPktRly1ySGmCYS1IDDHNJaoBhLkkN8AvQ2WTefK8+kTQthvls8uBqr1CRNC2eZpGkBhjmktQAw1ySGmCYS1IDDHNJaoBhLkkNMMwlqQGGuSQ1wDCXpAYY5pLUAMNckhpgmEtSAwxzSWqAYS5JDTDMJakBhrkkNcAwl6QGGObSGrDtkqVExECPbZcsHXa5apD/Nk5aA2684Xr/5Z+Gypa5JDXAMJekBhjmktQAw1ySGmCYS1IDDHNJaoBhLkkNMMwlqQGGuSQ1wDCXpAYY5pLUAMNcDzdv/sA3jJq/3oKBh53qDaa8cZU0Nd5oSw/34Oop3TBq0GFHhh+UN66SpsaWuSQ1wDCXpAYY5pLUAMNckhpgmEuN8Aqguc2rWaRGeAXQ3GbLXJIaYJhLUgMMc0lqgGEuSQ0wzDWnTOWKj6nce0Zrr1auAvJqFs0pU73iw6tD2tfKVUC2zCWpAYa5JDVgKGEeEcsi4kcR8eOIOHIYNUhSS2Y8zCNiHeAfgb2BnYCDImKnma5DkloyjJb5s4EfZ+a1mflL4AvAvkOoQ5KaEZk5szOM2B9YlplvqK8PBp6TmX84arjDgcPryycCP5rRQgezELht2EWMYzbXBrO7Pmubvtlc3zBruy0zl/U5g2FcmjjWRbn/6x0lM48Fju2/nOmLiBWZucuw6xjLbK4NZnd91jZ9s7m+2VzbmjCM0yw3AEs6r7cDbhxCHZLUjGGE+XeBHSPiMRGxHnAgcPoQ6pCkZsz4aZbMXB0Rfwh8HVgH+ExmXjnTdawhs/k00GyuDWZ3fdY2fbO5vtlc2yM241+ASpLWPH8BKkkNMMwlqQGG+Tgmu+VARLwmIi6vjwsi4qmdfisj4ocRcVlErBhCbXtExJ11/pdFxHsGHXcGantnp64rIuKBiNiy9ut7vX0mIm6NiCvG6R8R8Xe19ssj4hmDLtcM1DbM/W2y2oa2vw1Y39D2uRmVmT5GPShfzP4n8FhgPeAHwE6jhtkN2KI+3xu4uNNvJbBwiLXtAZwxnXH7rm3U8C8DvjkT661O/zeBZwBXjNN/H+BfKL+F2HVkm/a93gasbSj724C1DWV/G7S+Ye5zM/mwZT62SW85kJkXZObt9eVFlOvlZ0VtPY3bx/QPAk5cg/OfUGaeD/xsgkH2BY7P4iJg84hYzAzcgmKy2oa4vw2y3sYzI7fumGJ9M7rPzSTDfGzbAtd3Xt9Qu43nMEqLbkQC34iI79XbEgyjtudGxA8i4l8iYucpjtt3bUTEhsAy4Mudzn2ut0GMV3/f622qZnJ/G9Qw9rcpmaX73Brjfxoa20C3HACIiN+iHFy7dzo/LzNvjIitgLMj4uraepip2i4Fts/MeyJiH+ArwI4Djtt3bSNeBnwnM7stqj7X2yDGq7/v9TawIexvgxjW/jZVs3GfW2NsmY9toFsORMT/AT4N7JuZPx3pnpk31r+3AqdSPm7OWG2ZeVdm3lOfnwmsGxELBxm379o6DmTUx92e19sgxqt/VtyCYkj726SGuL9N1Wzc59acYZ+0n40PyieWa4HH8NAXNzuPGmYp8GNgt1HdNwI26Ty/gHKXyJms7dE89IOwZwM/obSSJh2379rqcJtRznFuNFPrrTOfHRj/i7yX8PAvQC+ZynL1XNtQ9rcBaxvK/jZofcPe52bq4WmWMeQ4txyIiDfV/p8E3gM8Cvh4lP/OvjrLHdm2Bk6t3eYDn8/Ms2a4tv2BN0fEauDnwIFZ9theb6UwYG0ALwe+kZn3dkbvdb0BRMSJlCsvFkbEDcBfAut2ajuTckXLj4H7gN+baLlmuLah7G8D1jaU/W0K9cGQ9rmZ5M/5JakBnjOXpAYY5pLUAMNckhpgmEtSAwxzSWqAYa7eRcSZEbH5JMO8LyL2mub094iIM6ZV3FomIt497Bo0O3lponoT5QLeyMwHe57PHsA7MvOlfc5nNoiIezJz42HXodnHlrmmLSLeXu8PfUVEHFG77RARV0XExyn37FhS7xm9sPb/i4i4OiLOjogTI+IdtftxEbF/fb4yIt4bEZfWe00/qXZ/dr2X9/fr3ydOUt+GEXFylHuAnxQRF0fELrXfiyLiwjqPL0bExpPMe6Mo983+bp3/mHf/i4g/qeP9ICKOqt2eFhEX1TpOjYgtavfzOvUsjIiV9fmhEXFKRJwVEddExIdr96OADaLce/tztaav1XldERGvnuamVAuG/RNUH2vnA3gm8EPKz6A3Bq4Enk75WfWDwK6dYVcCC4FdgMuADYBNgGsoLWqA44D9O8O/tT5/C/Dp+nxTYH59vhfw5fp8D8a+n/Y7gGPq86cAq2sNC4HzqT/tBt4FvGeSeX8QeG19vjnwH3R+Gl677035SfiG9fWW9e/lwAvq8/cBR9fn5wG71OcLgZX1+aGUn8FvBiwArgOW1H73dOb3SuBTndebDXu/8DG8hy1zTdfuwKmZeW+WmyydAjy/9rsuy/3AxxrntMz8eWbeDXx1gumfUv9+j/IGASXcvhjlP8p8FNh5jPFGz+8LAJl5BSVUodx3ZSfgOxFxGXAIsP0k834RcGQd/jxKyC4dNb+9gH/OzPvqPH8WEZsBm2fmt+owyyn/TGEy52TmnZn5C+DfR9U34ofAXhHxoYh4fmbeOcB01SjvzaLpGuv2piPuHaf7ROOMdn/9+wAP7ad/DZybmS+PiB0ooTqR8eYXwNmZedAU5h3AKzPzR5PMbypfQq3moVOdC8apYXQdv5aZ/xERz6TcT+ZvIuIbmfm+KcxfDbFlruk6H9ivnpfeiHIjo3+bZJxvAy+LiAX1HPVLpjjPzYD/rs8PHWD4bwOvAoiInYDfqN0vAp4XEY+v/TaMiCdMMq2vA2+tX+oSEU8fY5hvAK+P8k8QiIgta2v59ogY+dRyMDDSSl9JOV0F5WZVg/hVRKxbp78NcF9mngB8hPKv0zRH2TLXtGTmpRFxHHBJ7fTpzPx+bTGPN853I+J0yq1QrwNWAFM5NfBhYHlEvB345gDDf7wOfznwfcppljszc1VEHAqcGBHr12H/nHIefDx/DRwNXF4DfSXwsKtnMvOsiHgasCIifkm5C+O7KadxPllD/lrq3RgpAXxyRBw84PIAHFtruBQ4Hvh/EfEg8CvgzQNOQw3y0kTNqIjYOMt/pNmQ0ro/PDMv7Wle6wDrZuYvIuJxwDnAE7L8P0qpKbbMNdOOrac8FgDL+wryakPg3HpaIoA3G+RqlS1zSWqAX4BKUgMMc0lqgGEuSQ0wzCWpAYa5JDXgfwDcyHyU5A3DKgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "### output info of genes that after perturbation are 0\n",
    "to_zero=analyse.knocked_gene_info[analyse.knocked_gene_info['knocked']==0]\n",
    "sns.displot(to_zero['original'])\n",
    "plt.title('Original expression level of genes after perturbation is 0')\n",
    "plt.xlabel('original gene counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "pandas2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "anndata=sc.read_h5ad('/home/rui/NFS/mount/Genome-scale_Perturb-seq/rpe1_raw_singlecell_01.h5ad')\n",
    "exp=pd.DataFrame(anndata.X, columns=anndata.var.index, index=anndata.obs.index).T\n",
    "rename_dict = dict(zip(anndata.var.index, anndata.var['gene_name']))\n",
    "rename_dict['ENSG00000284024'] = 'MSANTD7'\n",
    "exp.rename(index=rename_dict, inplace=True)\n",
    "gene_list=list(set(anndata.obs['gene']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "for gene in gene_list:\n",
    "    exp_filtered = pd.DataFrame(exp[(anndata.obs[anndata.obs['gene'] == gene]).index])\n",
    "\n",
    "\n",
    "r_df_cdata = pandas2ri.py2rpy(anndata.obs.loc[exp_filtered.T.index]['gene'])\n",
    "r_df_fdata = pandas2ri.py2rpy(pd.DataFrame(exp_filtered.index))\n",
    "\n",
    "# exp_filtered=exp_filtered.reset_index(drop=True)\n",
    "# exp_filtered.columns=range(len(exp_filtered.columns))\n",
    "r_df_exp = pandas2ri.py2rpy(exp_filtered)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "r_code = \"\"\"\"\n",
    "library(MAST)\n",
    "mast_object <- FromMatrix(r_df_exp, r_df_cdata, r_df_fdata)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# r(r_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found existing installation: tf-nightly 2.14.0.dev20230723\n",
      "Uninstalling tf-nightly-2.14.0.dev20230723:\n",
      "  Would remove:\n",
      "    /home/rui/software/anaconda3/bin/estimator_ckpt_converter\n",
      "    /home/rui/software/anaconda3/bin/import_pb_to_tensorboard\n",
      "    /home/rui/software/anaconda3/bin/saved_model_cli\n",
      "    /home/rui/software/anaconda3/bin/tf_upgrade_v2\n",
      "    /home/rui/software/anaconda3/bin/tflite_convert\n",
      "    /home/rui/software/anaconda3/bin/toco\n",
      "    /home/rui/software/anaconda3/bin/toco_from_protos\n",
      "    /home/rui/software/anaconda3/lib/python3.9/site-packages/tensorflow/*\n",
      "    /home/rui/software/anaconda3/lib/python3.9/site-packages/tf_nightly-2.14.0.dev20230723.dist-info/*\n",
      "Proceed (Y/n)? ^C\n",
      "\u001b[31mERROR: Operation cancelled by user\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "! pip uninstall tf-nightly -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
